File: src/120-triangle/typescript/triangle.ts
// function minimumTotal(triangle: number[][]): number { //solution based on DP with O(1)
//     if (triangle.length === 1)
//         return triangle[0][0];
//     for (let level: number = triangle.length - 2; level >= 0; level--) {
//         for (let col: number = 0; col < triangle[level].length; col++)
//             triangle[level][col] += Math.min(triangle[level + 1][col], triangle[level + 1][col + 1]);
//     }
    
//     return triangle[0][0];
// };

function minimumTotal(triangle: number[][]): number { //solution based on DP with extra space
    const ans: number[] = new Array(triangle.length + 1).fill(0);

    for (let i = triangle.length - 1; i >= 0; i--) {
        for (let j = 0; j < triangle[i].length; j++)
            ans[j] = Math.min(ans[j], ans[j + 1]) + triangle[i][j];
    }

    return ans[0];
};

File: src/120-triangle/java/Triangle.java
import java.util.List;

class Triangle {
    //Dynamic programming solution
    public int minimumTotal(List<List<Integer>> triangle) {
        int[] ans = new int[triangle.size()+1];
        for(int i=triangle.size()-1;i>=0;i--){
            for(int j=0;j<triangle.get(i).size();j++)
                ans[j] = Math.min( ans[j],ans[j+1] ) + triangle.get(i).get(j);
        }
        return ans[0];
    }
    //same algo without extra space (slower time complexity)
    // public int minimumTotal(List<List<Integer>> triangle) {
    //     for(int i = triangle.size()-2 ; i>=0; i--) {
    //         for(int j=0; j<=i; j++)
    //             triangle.get(i).set(j, triangle.get(i).get(j) + Math.min( triangle.get(i+1).get(j), triangle.get(i+1).get(j+1)));
    //     }
    //     return triangle.get(0).get(0);
    // }
}